File: app/src/main/java/com/supercilex/robotscouter/ui/scout/viewholder/ScoutViewHolderBase.java
package com.supercilex.robotscouter.ui.scout.viewholder;

import android.annotation.SuppressLint;
import android.support.annotation.CallSuper;
import android.support.v4.app.FragmentManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.SimpleItemAnimator;
import android.text.TextUtils;
import android.view.View;
import android.widget.TextView;

import com.supercilex.robotscouter.R;
import com.supercilex.robotscouter.data.model.metrics.ScoutMetric;

public abstract class ScoutViewHolderBase<TMetric, VView extends TextView> extends RecyclerView.ViewHolder {
    protected final VView mName;
    protected ScoutMetric<TMetric> mMetric;
    protected FragmentManager mManager;

    private SimpleItemAnimator mAnimator;

    @SuppressLint("WrongViewCast")
    public ScoutViewHolderBase(View itemView) {
        super(itemView);
        //noinspection unchecked
        mName = (VView) itemView.findViewById(R.id.name);
    }

    public final void bind(ScoutMetric<TMetric> metric,
                           FragmentManager manager,
                           SimpleItemAnimator animator) {
        mMetric = metric;
        mManager = manager;
        mAnimator = animator;

        bind();
    }

    @CallSuper
    protected void bind() {
        mName.setText(mMetric.getName());
    }

    protected void updateMetricName(String name) {
        if (!TextUtils.equals(mMetric.getName(), name)) {
            disableAnimations();
            mMetric.updateName(name);
        }
    }

    protected void updateMetricValue(TMetric value) {
        if (!value.equals(mMetric.getValue())) {
            disableAnimations();
            mMetric.updateValue(value);
        }
    }

    protected void disableAnimations() {
        mAnimator.setSupportsChangeAnimations(false);
    }

    @Override
    public String toString() {
        return mMetric.toString();
    }
}

File: app/src/main/java/com/supercilex/robotscouter/ui/scout/viewholder/ScoutViewHolderBase.kt
package com.supercilex.robotscouter.ui.scout.viewholder

import android.support.annotation.CallSuper
import android.support.v4.app.FragmentManager
import android.support.v7.widget.RecyclerView
import android.support.v7.widget.SimpleItemAnimator
import android.text.TextUtils
import android.view.View
import android.widget.TextView
import com.supercilex.robotscouter.R
import com.supercilex.robotscouter.data.model.metrics.ScoutMetric

abstract class ScoutViewHolderBase<TMetric, out VView : TextView>(itemView: View) :
        RecyclerView.ViewHolder(itemView) {
    @Suppress("UNCHECKED_CAST")
    protected val mName: VView = itemView.findViewById(R.id.name) as VView
    protected lateinit var mMetric: ScoutMetric<TMetric>
    protected lateinit var mManager: FragmentManager
    private lateinit var mAnimator: SimpleItemAnimator

    fun bind(metric: ScoutMetric<TMetric>, manager: FragmentManager, animator: SimpleItemAnimator) {
        mMetric = metric
        mManager = manager
        mAnimator = animator

        bind()
    }

    @CallSuper
    protected open fun bind() {
        mName.text = mMetric.name
    }

    protected fun updateMetricName(name: String) {
        if (!TextUtils.equals(mMetric.name, name)) {
            disableAnimations()
            mMetric.updateName(name)
        }
    }

    protected fun updateMetricValue(value: TMetric) {
        if (value != mMetric.value) {
            disableAnimations()
            mMetric.updateValue(value)
        }
    }

    protected fun disableAnimations() {
        mAnimator.supportsChangeAnimations = false
    }

    override fun toString() = mMetric.toString()
}
